src/document/utils/flatten-fill.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * @typedef {import("../builders.js").Doc} Doc
+ */
+
+import { DOC_TYPE_ARRAY, DOC_TYPE_FILL } from "../constants.js";
+import { getDocParts } from "../utils.js";
+import { fillBuilder } from "./fill-builder.js";
+import getDocType from "./get-doc-type.js";
+
+/**
+ * @param {Doc} doc
+ * @returns {Doc}
+ */
+function flattenFill(doc) {
+  if (![DOC_TYPE_ARRAY, DOC_TYPE_FILL].includes(getDocType(doc))) {
+    throw new Error("Expected doc to be fill()");
+  }
+  const builder = fillBuilder();
+
+  function drain(doc) {
+    switch (getDocType(doc)) {
+      case DOC_TYPE_FILL: {
+        const childParts = getDocParts(doc);
+        for (const [j, childPart] of childParts.entries()) {
+          if (j % 2 === 1) {
+            builder.appendLine(childPart);
+            continue;
+          }
+          drain(childPart);
+        }
+        break;
+      }
+      case DOC_TYPE_ARRAY:
+        for (const part of getDocParts(doc)) {
+          drain(part);
+        }
+        break;
+      default:
+        builder.append(doc);
+        break;
+    }
+  }
+
+  drain(doc);
+
+  return builder.build();
```

Comment:
Will something like this work?

```js
  const parts = []
  
  // Check last part / parts.index is ODD-EVEN
  switch (getDocType(doc)) {
      case DOC_TYPE_FILL: {
        // concat or turn the last element to array and push the first
        // push the rest
        break;
      }
      default:
        builder.append(doc);
        break;
    }

return fill(parts)
```

Reply:
@fisker 
I'm sorry but I don't understand what does it mean / aim at.
Do you mean
- it can be faster?
- it can be more readable?
- prefer cleaner doc? (e.g. prefer `fill("a", line, "b")` to `fill(["a"], line, ["b"])`)
- recursion should be avoided?
- fillBuilder looks over-engineering?

I'm not sure whether I understand the pseudo code or not but possibly following case can be a counter-example.
input: `["a", fill(["b", line, "c", line]), "b"]`
expected output: `fill(join(line, [["a", "b"], ["c"], ["b"]]))`

Reply:
It's simply, so we don't need `fillBuilder`. Can you find other places that we want a "flat" fill?

Reply:
@fisker 
Does [866b257](https://github.com/prettier/prettier/pull/15909/commits/866b257e4719b3ab1c5f27f7c78b731a3ca62f49) meet your request?